### ISSUE TITLE
Problem due to overloading `no-confirm` for "save"

### ISSUE THREAD

**Issue by mentalisttraceur** [protesilaos/denote#299]
I noticed that 2.3.0 overloaded the `no-confirm` parameter of `denote-rewrite-front-matter` to both turn off confirmation and turn on unconditional saving.

I appreciate that this was probably to help users like me who want both of these features, so thank you for that.

However, to me it's important to be able to control these two behaviors independently. Just because I don't want to be prompted doesn't mean I want to risk unrelated unwanted saving/clobberring, and if my checks report it's not safe to save (as described in #273) that still doesn't mean I want to be prompted about the frontmatter diff.

(I also don't think it's just me. These are two qualitatively different kinds and severity of (un)safety. Combining them in one boolean collapses their individual trade-offs to the worst of both: you either get prompted to confirm the in-buffer change just to not overwrite/save unrelated changes, or you have to pay the risk of the latter just to not deal with the former.)

Since neither of the new combined behaviors of `no-confirm` is ever the right thing for me, I am now back to injecting advice (personally I'm fine with this, almost every package I use I end up advising something, but I think it's a good symptom / illustration that two independent booleans got ORed into one here):

```lisp
    (defun my-denote-rewrite-front-matter-wrapper (path title tags type)
        (let* ((buffers (buffer-list))
               (buffer  (find-file-noselect path))
               (was-already-open    (memq buffer buffers))
               (had-unsaved-changes (when was-already-open
                                        (refresh-modified-state buffer)
                                        (buffer-modified-p buffer))))
            (with-advice ('y-or-n-p :override 'always)
                (denote-rewrite-front-matter path title tags type))
            (unless had-unsaved-changes
                (with-current-buffer buffer
                    (save-buffer)))
            (unless was-already-open
                (kill-buffer buffer))))
```

Personally, I would recommend

1. return `no-confirm` to its uncoupled behavior of just controlling the prompt;

2. one of either

    * add the additional optional save parameter that was mentioned in #273,
    * (if you're not already using the return value for anything) just return non-nil if you did the rewrite and `nil` otherwise (maybe it already ends up doing this, make it explicitly part of the API if so) or
    * make possible to reuse the front matter rewrite prompt, same as we can use the rename prompt

    Any one of these three allows people who *do* want the prompt to have the save conditional on the prompt. Contrast in my wrapper above, where if I did want to re-enable the prompt, my code couldn't easily hold back from saving even if I hit `n`.

P.S. thanks for all the work in 2.3.0! I can see the consideration for users like me, and I was able to clean out a couple hacks from my config thanks to what you've added. (re: GNU copyright papers, still waiting on reply)

**Comment by mentalisttraceur** [comment]
Thinking about it a bit more, I think pulling save into the lowest-level-but-still-public-API rewrite function is probably not the right play - better to just empower composition.

So for point 2 above, I now recommend the return value approach - just allowing the caller to know after the call if the rewrite happened is more flexible than pulling the save into the rewrite.

**Comment by jeanphilippegg** [comment]
You are right. The coupling is not ideal.

In my fork of the code (not announced yet elsewhere), I have made the options `denote-rename-confirmations` and `denote-save-buffers` control how the renaming commands confirm and save buffers. All renaming commands will respect those variables. None will have any optional parameters anymore regarding that.

`denote-rename-confirmations` will default to `'(add-front-matter rewrite-front-matter modify-file-name)` and `denote-save-buffers` to `nil`.

In the release notes for version 2.3.0, Prot has expressed a wish to wait until late April for new features.

I plan on making the pull requests to include everything that can be included back from my fork into Denote.

But first, there will be 2 other pull requests that will come before this one: I will first add `denote-forced-*` variables to help in customizing denote commands and then I will submit a refactor of all renaming commands to remove all code duplication and enhance their flexibility. It will be very easy for developers as well as users to define their own custom renaming commands.

It will be very easy to have separate renaming commands for all file name components (or multiple).

This is all done already in my fork. I will make it public this weekend, but I will do the pull requests in May!

**Comment by protesilaos** [comment]
Yeah, the idea is to wait a bit for new features or more fundamental changes so that we can fix obvious bugs. Once we are sure that users are okay with what they have, we resume the work. A more fine-grained control over these confirmations is a good idea.

**Comment by mentalisttraceur** [comment]
Yep, I saw that in the release notes and totally respect that - I'm raising this now because this could be seen as a bug, and point 1 of my proposal as the fix.

Consider: if I hadn't read the changelog and checked the docstrings of functions I use, I wouldn't have realized that code which I wrote+tested+trusted to only save when deemed safe by my checks would now suddenly start saving files in other situations as well.

*Personally*, I don't mind - I already protected myself from this by proactively thoroughly checking the change log\* and immediately adjusting my code. Cool, no data loss happened, no bad feelings.

But I think users would have been justified in expecting that the meaning/effect of a parameter named "no confirm" in a public API function wouldn't get promoted to "no confirm and save".

On the other hand, if you've got one or two users who prefer the new behavior, maybe that's more representative of a typical user. So I'm not saying that you definitely should treat this as a bug, I'm just suggesting to think about it.

\* Speaking of, thank you very much for that thorough change log! I know those are really hard to write even in basic form, and yours is both thorough and has an *impressive* degree of polish.
